core/drivers/stm32_gpio.c

```diff
@@ -606,6 +725,178 @@ static bool bank_is_registered(const void *fdt, int node)
+	io_clrsetbits32(bank->base + GPIO_SECR_OFFSET, GPIO_SECCFGR_MASK,
+			bank->rif_cfg->sec_conf[0]);
+
+	if (!bank->is_tdcid)
+		goto out;
+
+	for (i = 0; i < bank->ngpios; i++) {
+		if (!(BIT(i) & bank->rif_cfg->access_mask[0]))
+			continue;
+
+		io_clrsetbits32(bank->base + GPIO_CIDCFGR(i),
+				GPIO_CIDCFGR_CONF_MASK,
+				bank->rif_cfg->cid_confs[i]);
+
+		cidcfgr = io_read32(bank->base + GPIO_CIDCFGR(i));
+
+		/*
+		 * Take semaphore if the resource is in semaphore mode
+		 * and secured.
+		 */
+		if (!stm32_rif_semaphore_enabled_and_ok(cidcfgr, RIF_CID1) ||
+		    !(io_read32(bank->base + GPIO_SECR_OFFSET) & BIT(i))) {
+			res = stm32_rif_release_semaphore(bank->base +
+				GPIO_SEMCR(i),
+				GPIO_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Could not release semaphore for pin%u",
+				     i);
+				goto out;
+			}
+		} else {
+			res = stm32_rif_acquire_semaphore(bank->base +
+				GPIO_SEMCR(i),
+				GPIO_MAX_CID_SUPPORTED);
+			if (res) {
+				EMSG("Could not acquire semaphore for pin%u",
+				     i);
+				goto out;
+			}
+		}
+	}
+
+	/*
+	 * Lock RIF configuration if configured. This cannot be undone until
+	 * next reset.
+	 */
+	io_setbits32(bank->base + GPIO_RCFGLOCKR_OFFSET,
+		     bank->rif_cfg->lock_conf[0]);
+
+	if (IS_ENABLED(CFG_TEE_CORE_DEBUG)) {
+		/* Check that RIF config are applied, panic otherwise */
+		if ((io_read32(bank->base + GPIO_PRIVCFGR_OFFSET) &
+		     bank->rif_cfg->access_mask[0]) !=
+		    bank->rif_cfg->priv_conf[0]) {
+			EMSG("GPIO bank priv conf (ID: %u) is incorrect",
+			     bank->bank_id);
+			panic();
+		}
+
+		if ((io_read32(bank->base + GPIO_SECR_OFFSET) &
+		     bank->rif_cfg->access_mask[0]) !=
+		    bank->rif_cfg->sec_conf[0]) {
+			EMSG("GPIO bank sec conf (ID: %u) is incorrect",
+			     bank->bank_id);
+			panic();
+		}
+	}
+
+	res = TEE_SUCCESS;
+out:
+	clk_disable(bank->clock);
+
+	return res;
+}
+
+static void stm32_gpio_save_rif_config(struct stm32_gpio_bank *bank)
+{
+	size_t i = 0;
+
+	for (i = 0; i < bank->ngpios; i++)
+		bank->rif_cfg->cid_confs[i] = io_read32(bank->base +
+							 GPIO_CIDCFGR(i));
+
+	bank->rif_cfg->priv_conf[0] = io_read32(bank->base +
+						GPIO_PRIVCFGR_OFFSET);
+	bank->rif_cfg->sec_conf[0] = io_read32(bank->base +
+					       GPIO_SECR_OFFSET);
+	bank->rif_cfg->lock_conf[0] = io_read32(bank->base +
+						GPIO_RCFGLOCKR_OFFSET);
+}
+
+static void stm32_parse_gpio_rif_conf(struct stm32_gpio_bank *bank,
+				      const void *fdt, int node)
+{
+	unsigned int i = 0;
+	unsigned int nb_rif_conf = 0;
+	int lenp = 0;
+	const fdt32_t *cuint = NULL;
+
+	cuint = fdt_getprop(fdt, node, "st,protreg", &lenp);
+	if (!cuint) {
+		DMSG("No RIF configuration available");
+		return;
+	}
+
+	bank->rif_cfg = calloc(1, sizeof(*bank->rif_cfg));
+	if (!bank->rif_cfg)
+		panic();
+
+	bank->rif_cfg->sec_conf = calloc(1, sizeof(uint32_t));
+	if (!bank->rif_cfg->sec_conf)
+		panic();
+
+	nb_rif_conf = MIN((unsigned int)(lenp / sizeof(uint32_t)),
+			  bank->ngpios);
```

Comment:
If think we should rather report an error if `lenp / sizeof(uint32_t) > bank->ngpios`.

Reply:
I put an assert but let me know if an error case handling feels better to you